lib/token/symmetric.go

```
@@ -0,0 +1,194 @@
+func ReadOrGenerateSymmetricKey(path string, size int) SymmetricSetter {
+	if size == 0 {
+		size = 256
+	}
+
+	return func(be *SymmetricEncoder) error {
+		var err error
+		var key []byte
+		if path != "" {
+			key, err = ioutil.ReadFile(path)
+		}
+		if err != nil || len(key) <= 0 || len(key)*8 != size {
+			if err := WithGeneratedSymmetricKey(size)(be); err != nil {
+				return err
+			}
+			if path != "" {
+				err = ioutil.WriteFile(path, be.key, 0600)
+			}
+		} else {
+			err = UseSymmetricKey(key)(be)
+		}
+		return err
+	}
+}
+
+// NewSymmetricEncoder creates a new encoder using AES in GCM mode as a symmetric cipher.
+//
+// A typical way to use the encoder would be:
+//
+//    rng := rand.New(srand.Source)  // using github.com/enkit/lib/srand library.
+//    ...
+//    be, err := NewSymmetricEncoder(rng, ReadOrGenerateSymmetricKey("/etc/keys/connect.key", 0))
+//    if err != nil ...
+//
+// Or:
+//
+//    key, err := GenerateSymmetricKey(rng, 0)
+//    if err != nil ...
+//
+//    be, err := NewSymmetricEncoder(rng, UseSymmetricKey(key))
+//    ...
+//
+// followed by calls to Encode() and Decode().
+func NewSymmetricEncoder(rng *rand.Rand, setters ...SymmetricSetter) (*SymmetricEncoder, error) {
+	be := &SymmetricEncoder{rng: rng}
+	for _, setter := range setters {
+		if err := setter(be); err != nil {
+			return nil, err
+		}
+	}
+
+	if len(be.key) <= 0 {
+		return nil, fmt.Errorf("No key set")
+	}
+
+	block, err := aes.NewCipher(be.key)
+	if err != nil {
+		return nil, err
+	}
+
+	be.cipher, err = cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
+
+	return be, nil
+}
+
+// SymmetricCreator is a CryptoFactory suitable for use with NewMultiKeyCryptoEncoder().
+var SymmetricCreator CryptoFactory = func(rng *rand.Rand, key []byte) (BinaryEncoder, []byte, error) {
+	if key != nil {
+		be, err := NewSymmetricEncoder(rng, UseSymmetricKey(key))
+		return be, key, err
+	}
+
+	key, err := GenerateSymmetricKey(rng, 0)
```

I don't like this behaviour. I feel like it shouldn't default to random if not set. I don't know feels like it could lead to a lot of misleading tests for applications.

---

In this specific function, CryptoFactory, it has to be a random key. This is used by the MultiCrypto encoder, which needs "something" to a) provide a random/unguessable/secret key to use to encrypt the message once, and then 2) an encoder capable of encrypting that data.

Then, this unguessable key is encrypted with each key that should have access to the message in turn, so any of those *other* keys can be used for crypto purposes.

The other encoders default to use a random key as a) it makes development simpler (start server, it works), b) it is the safest (eg, no secret stored on disk), and c) is "acceptable" in many deployments (single backend, restarted rarely, tokens are valid until timeout or until backend is restarted).

But also orthogonal to this PR?